### PR TITLE
Add `UNet3DTrainer` (base class for training UNet3D in PyTorch CUDA and PyTorch/XLA)

### LINKED ISSUE #8: Refactor `training` module for readability and extensibility with `torch_xla`
We hope to refactor the `training` module into a more readable and extensible `trainer` package.

The `trainer` package uses polymorphism (the base class is `UNet3DTrainer` & the concrete sub-classes are `CUDATrainer` and `XLATrainer`) to toggle between training with CUDA and training with PT-XLA during runtime.

For more details, see b/224290413

### PR DESCRIPTION
**Pull request by thisisalbertliang**
Resolve #8 (partially)

Related to
- `XLATrainer` PR: https://github.com/thisisalbertliang/training/pull/10
- `CUDATrainer` PR: https://github.com/thisisalbertliang/training/pull/11

We hope to refactor the `training` module into a more readable and extensible `trainer` package.

The `trainer` package uses polymorphism (the base class is `UNet3DTrainer` & the concrete sub-classes are `CUDATrainer` and `XLATrainer`) to toggle between training with CUDA and training with PT-XLA during runtime.

For more details, see b/224290413

### CODE REVIEW

File: image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+from model.unet3d import Unet3D
+from runtime.distributed.distributed_utils import get_world_size, reduce_tensor
+from runtime.inference import evaluate
+from runtime.logging import CONSTANTS, mllog_end, mllog_event, mllog_start
+from torch.nn import Parameter
+from torch.optim import SGD, Adam
+
+
+class UNet3DTrainer(ABC):
+    """Base class for training UNet3D in PyTorch CUDA and PyTorch/XLA"""
```
Comment (miladm) [inline image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py]: I'd not specify CUDA and PyTorc/XLA. You are building this to be widely extensible to any backend framework. 

Suggestion: "Base class for training UNet3D in PyTorch"

File: image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+        self.callbacks = callbacks
+        self.score_fn = score_fn
+        self.device = device
+        self.train_loader = train_loader
+        self.val_loader = val_loader
+        self.model: Unet3D = model
+        self.loss_fn: torch.nn.Module = loss_fn
+
+        # move model and loss_fn to device
+        self.model.to(device)
```
Comment (miladm) [inline image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py]: nit: `self.device`

File: image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        self.loss_fn.to(device)
```
Comment (miladm) [inline image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py]: ditto

File: image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+                image, label = batch
+                image, label = image.to(self.device), label.to(self.device)
+
+                for callback in self.callbacks:
+                    callback.on_batch_start()
+
+                loss_value = self.forward_pass(image, label)
+                self.backward_pass(iteration, loss_value)
+
+                loss_value = reduce_tensor(loss_value).detach().cpu().numpy()
```
Comment (miladm) [inline image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py:105]: Do we know if this operation is necessary for a XLA device? How about a TPU device?

See this for [reference](https://github.com/pytorch/pytorch/blob/master/torch/multiprocessing/reductions.py#L139-L276)

**Review by miladm — CHANGES_REQUESTED** [review]
Thanks @thisisalbertliang!

I add a few comments below.

File: image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py (hunk 1 above)
Comment (thisisalbertliang) [reply image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py]: Thanks for the suggestion. I just changed the docstring to `"""Base class for training UNet3D in PyTorch"""`

File: image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py (hunk 2 above)
Comment (thisisalbertliang) [reply image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py]: done.

File: image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py (hunk 3 above)
Comment (thisisalbertliang) [reply image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py]: done.

File: image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py (hunk 4 above)
Comment (thisisalbertliang) [reply image_segmentation/pytorch/runtime/trainer/unet3d_trainer.py:105]: Thanks for bringing this up.

I just confirmed that this all-reduce op is indeed necessary. I modified the train loop to print the loss value before & after the all-reduce op:
```
print(f"ALBERT DEBUG [*** RANK {get_rank()} ***]: loss_value BEFORE reduce: {loss_value}")
loss_value = reduce_tensor(loss_value).detach().cpu().numpy()
print(f"ALBERT DEBUG [*** RANK {get_rank()} ***] : loss_value AFTER reduce: {loss_value}")
```
and ran the code on TPU v3-8. This is the output log I got for one iteration:
```
ALBERT DEBUG [*** RANK 6 ***]: loss_value BEFORE reduce: 1.0074005126953125
ALBERT DEBUG [*** RANK 2 ***]: loss_value BEFORE reduce: 1.009411334991455
ALBERT DEBUG [*** RANK 3 ***]: loss_value BEFORE reduce: 0.9671453237533569
ALBERT DEBUG [*** RANK 0 ***]: loss_value BEFORE reduce: 1.020106554031372
ALBERT DEBUG [*** RANK 1 ***]: loss_value BEFORE reduce: 0.993670642375946
ALBERT DEBUG [*** RANK 4 ***]: loss_value BEFORE reduce: 0.9977484345436096
ALBERT DEBUG [*** RANK 5 ***]: loss_value BEFORE reduce: 1.002981424331665
ALBERT DEBUG [*** RANK 7 ***]: loss_value BEFORE reduce: 0.9892218112945557
ALBERT DEBUG [*** RANK 1 ***] : loss_value AFTER reduce: 0.9984608888626099
ALBERT DEBUG [*** RANK 6 ***] : loss_value AFTER reduce: 0.9984608888626099
ALBERT DEBUG [*** RANK 7 ***] : loss_value AFTER reduce: 0.9984608888626099
ALBERT DEBUG [*** RANK 2 ***] : loss_value AFTER reduce: 0.9984608888626099
ALBERT DEBUG [*** RANK 5 ***] : loss_value AFTER reduce: 0.9984608888626099
ALBERT DEBUG [*** RANK 3 ***] : loss_value AFTER reduce: 0.9984608888626099
ALBERT DEBUG [*** RANK 4 ***] : loss_value AFTER reduce: 0.9984608888626099
ALBERT DEBUG [*** RANK 0 ***] : loss_value AFTER reduce: 0.9984608888626099
```

As shown above, the all-reduce op is indeed necessary for aggregating the loss value.

**Review by will-cromar — APPROVED** [review]

**Review by miladm — APPROVED** [review]
LGTM

Thanks @thisisalbertliang!
